realm/realm-library/src/main/cpp/io_realm_RealmFunctions.cpp

```diff
@@ -18,19 +18,40 @@
 
 #include "util.hpp"
 #include "jni_util/bson_util.hpp"
+#include "java_network_transport.hpp"
+#include "object-store/src/sync/app.hpp"
 
 using namespace realm;
+using namespace realm::app;
 using namespace realm::jni_util;
 
-// FIXME This is just a basic round trip test for passing bson back and forth. Proper implementation
-//  will come with actual Function implementation.
-JNIEXPORT jstring JNICALL Java_io_realm_RealmFunctions_nativeCallFunction
-        (JNIEnv* env, jclass, jstring j_args) {
+static std::function<jobject(JNIEnv*, Optional<bson::Bson> )> response_mapper = [](JNIEnv* env, Optional<bson::Bson> response) {
+    if (response) {
+        return JniBsonProtocol::bson_to_jstring(env, *response);
+    } else {
+        // FIXME How to raise errors here
+        return to_jstring(env, "{}");
+    }
+};
+
+JNIEXPORT void JNICALL
+Java_io_realm_RealmFunctions_nativeCallFunction(JNIEnv* env, jclass , jlong j_app_ptr, jlong j_user_ptr, jstring j_name,
+                                                jstring j_args_json , jobject j_callback) {
     try {
-        bson::Bson bson = JniBsonProtocol::jstring_to_bson(env, j_args);
-        return JniBsonProtocol::bson_to_jstring(env, bson);
+        auto app = *reinterpret_cast<std::shared_ptr<App>*>(j_app_ptr);
+        auto user = *reinterpret_cast<std::shared_ptr<SyncUser>*>(j_user_ptr);
+
+        std::function<void(Optional<bson::Bson>, Optional<app::AppError>)> callback;
```

Comment:
I assume the reason for doing this in two lines is for readability? But there is a slight performance penalty as you are creating a dummy callback that is being thrown away straight away.

Reply:
👍